transaction/core/src/input_rules.rs

```diff
@@ -53,6 +69,112 @@ impl InputRules {
+        &self,
+        _block_version: BlockVersion,
+        tx: &Tx,
+    ) -> Result<(), InputRuleError> {
+        if let Some(partial_fill_change) = self.partial_fill_change.as_ref() {
+            // There is a partial fill change output. Let's try to unblind its amount.
+            let partial_fill_change_amount = partial_fill_change.reveal_amount()?;
+            // If the min fill value exceeds the fractional change, then the SCI is
+            // ill-formed
+            if self.min_partial_fill_value > partial_fill_change_amount.value {
+                return Err(InputRuleError::MinPartialFillValueExceedsPartialFillChange);
+            }
+
+            // Let's check if there is a corresponding fractional change output.
+            let fractional_change = tx
+                .prefix
+                .outputs
+                .iter()
+                .find(|x| partial_fill_change.tx_out.matches_ignoring_amount(x))
+                .ok_or(InputRuleError::MissingFractionalChangeOutput)?;
+            // Let's try to unblind its amount.
+            let fractional_change_amount =
+                try_reveal_amount(fractional_change, partial_fill_change.amount_shared_secret.as_ref())?;
+
+            // Check the bounds of the real change amount
+            if fractional_change_amount.token_id != partial_fill_change_amount.token_id {
+                return Err(InputRuleError::FractionalOutputTokenIdMismatch);
+            }
+            // Partial fill change amount - min fill value is an upper bound on how much
+            // can be returned in the fractional change output.
+            if fractional_change_amount.value > partial_fill_change_amount.value - self.min_partial_fill_value {
+                return Err(InputRuleError::FractionalChangeOutputAmountExceedsLimit);
+            }
+
+            // Compute fill-fraction num and denom. This is the fraction of the maximum
+            // possible trade that could have occurred which did occur. This
+            // calculation is used later when checking if real output amounts respected the
+            // fill fraction.
+            let fill_fraction_num =
+                (partial_fill_change_amount.value - fractional_change_amount.value) as u128;
+            let fill_fraction_denom = partial_fill_change_amount.value as u128;
+
+            // Verify partial_fill_outputs
+            for partial_fill_output in self.partial_fill_outputs.iter() {
+                // Try to unblind the partial fill output amount.
+                let partial_fill_output_amount = partial_fill_output.reveal_amount()?;
+                // Let's check if there is a corresponding fractional output.
+                let fractional_output = tx
+                    .prefix
+                    .outputs
+                    .iter()
+                    .find(|x| partial_fill_output.tx_out.matches_ignoring_amount(x))
+                    .ok_or(InputRuleError::MissingFractionalOutput)?;
```

Comment:
@garbageslam,
One question that came up talking with @samdealy yesterday
&iquest; What happens if the counterparty provides 2 outputs that match.  I.e. 20MOB back to themselves for the swap, copying that txout for another 20MOB back to themselves?

I know there was mention that certain keys could ever show up once in the blockchain, is that these txout keys?

Reply:
yeah that's exactly right :)

at the protocol level, consensus prevents two TxOut's from having the same TxOut public key, and a Tx is invalid if it has two TxOut's like that

Reply:
But I guess the other part is, the required_output and fractional_output generally don't go to the counterparty, they are created by the originator and go back to the originator. The part that the counterparty gets is the input underlying the SCI. So copying required or fractional outputs would just be giving more money to the originator, so the counterparty would have no reason to do that